data/transactions/logic/assembler.go

```diff
@@ -1428,10 +1433,13 @@ func typeStores(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes, e
 	if top < 0 {
 		return nil, nil, nil
 	}
+	var err error
 	for i := range pgm.scratchSpace {
-		// We can't know what slot stacktop is being stored in, but we can at least keep the slots that are the same type as stacktop
-		if pgm.scratchSpace[i] != pgm.stack[top] {
-			pgm.scratchSpace[i] = StackAny
+		// We can't know what slot stacktop is being stored in
+		// so we union it into all scratch slots
+		pgm.scratchSpace[i], err = pgm.scratchSpace[i].union(pgm.stack[top])
```

Comment:
cc @jannotti this is where we're doing the union now

Reply:
yeah, that seems right

If we wanted to make a unit test to show it working, is there any way, in this PR, to produce a bounded int? (does `int 5` even produce one?)

Reply:
oh, except this manipulates the scratch pace even when there's an error in union. that's seems bad, though I don't know what kind of errrors are possible.

Reply:
should be able to use the pseudo op `int` + a type* refine fn to produce a narrowed type 

https://github.com/barnjamin/go-algorand/pull/15

But I'll have to think about how exactly it could be tested

Reply:
https://github.com/algorand/go-algorand/pull/5130/commits/5f9b73a1af8eee45a1faa0e0ff225e76b4b5ee3f added a test here to show its behavior